pkg/proxy/endpoints.go

```diff
@@ -0,0 +1,231 @@
+type UpdateEndpointMapResult struct {
+	// HCEndpointsLocalIPSize maps an endpoints name to the length of its local IPs.
+	HCEndpointsLocalIPSize map[types.NamespacedName]int
+	// StaleServiceNames identifies if an endpoints service pair is stale.
+	StaleEndpoints map[EndpointServicePair]bool
+	// StaleServiceNames identifies if a service is stale.
+	StaleServiceNames map[ServicePortName]bool
+}
+
+// UpdateEndpointsMap updates endpointsMap base on the given changes.
+// The changes map is cleared after applying them.
+func UpdateEndpointsMap(endpointsMap EndpointsMap, changes *EndpointsChangeMap, hostname string) (result UpdateEndpointMapResult) {
+	result.StaleEndpoints = make(map[EndpointServicePair]bool)
+	result.StaleServiceNames = make(map[ServicePortName]bool)
+
+	func() {
+		endpointsMap.apply(changes, result.StaleEndpoints, result.StaleServiceNames)
+	}()
+
+	// TODO: If this will appear to be computationally expensive, consider
+	// computing this incrementally similarly to endpointsMap.
+	result.HCEndpointsLocalIPSize = make(map[types.NamespacedName]int)
+	localIPs := GetLocalIPs(endpointsMap)
+	for nsn, ips := range localIPs {
+		result.HCEndpointsLocalIPSize[nsn] = len(ips)
+	}
+
+	return result
+}
+
+// apply the changes to EndpointsMap and updates stale endpoints and service-endpoints pair. The `staleEndpoints` argument
+// is passed in to store the stale udp endpoints and `staleServiceNames` argument is passed in to store the stale udp service.
+func (endpointsMap EndpointsMap) apply(changes *EndpointsChangeMap, staleEndpoints map[EndpointServicePair]bool, staleServiceNames map[ServicePortName]bool) {
+	changes.lock.Lock()
+	defer changes.lock.Unlock()
+	for _, change := range changes.items {
+		endpointsMap.Unmerge(change.previous)
+		endpointsMap.Merge(change.current)
+		detectStaleConnections(change.previous, change.current, staleEndpoints, staleServiceNames)
+	}
+	changes.items = make(map[types.NamespacedName]*endpointsChange)
+}
+
+// Merge adds other ProxyEndpointsMap's elements to current ProxyEndpointsMap, if exists, update it, otherwise create new one.
+func (em EndpointsMap) Merge(other EndpointsMap) {
+	for svcPortName := range other {
+		em[svcPortName] = other[svcPortName]
+	}
+}
+
+// Unmerge deletes all other EndpointsMap's elements from current EndpointsMap.
+func (em EndpointsMap) Unmerge(other EndpointsMap) {
+	for svcPortName := range other {
+		delete(em, svcPortName)
+	}
+}
+
+// GetLocalIPs returns endpoints IPs if given endpoints is local.
+func GetLocalIPs(endpointsMap EndpointsMap) map[types.NamespacedName]sets.String {
+	localIPs := make(map[types.NamespacedName]sets.String)
+	for svcPortName := range endpointsMap {
+		for _, ep := range endpointsMap[svcPortName] {
+			if ep.IsLocal() {
+				nsn := svcPortName.NamespacedName
+				if localIPs[nsn] == nil {
+					localIPs[nsn] = sets.NewString()
+				}
+				localIPs[nsn].Insert(ep.IPPart()) // just the IP part
```

Comment:
I think you can remove the comment here since it is just reiterating what the code is clearly doing.

Reply:
It's fixed now.